commit c52e691c9c0e2876e29c795a46dca9439e07b516 (refs/changes/87/1487/2)
Author: Jordan Paige Hendricks <jordan.hendricks@joyent.com>
Date:   2017-02-10T01:40:18+00:00 (2 years, 8 months ago)
    
    MANTA-3150 muskie logging about shards is sometimes incorrect, and could also be more useful

diff --git a/lib/audit.js b/lib/audit.js
index 36bfa92..c4b2951 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -231,7 +231,8 @@ function auditLogger(options) {
             obj.objectId = req.metadata.objectId;
         }
         obj.sharksContacted = req.sharksContacted;
-        obj.shard = req.shard;
+        obj.baseShard = req.objectShard;
+        obj.dirShard = req.parentShard;
 
         if (req._timeToLastByte !== undefined &&
             req._totalBytes !== undefined) {
diff --git a/lib/common.js b/lib/common.js
index 8996f9f..eebccf4 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -372,6 +372,9 @@ function ensureParent(req, res, next) {
 }
 
 
+/*
+ * Resolves the object metadata and the parent metadata, if necessary.
+ */
 function getMetadata(req, res, next) {
     var log = req.log;
 
@@ -392,6 +395,11 @@ function getMetadata(req, res, next) {
                             metadata: md,
                             etag: (w || {})._etag
                         };
+
+                        if (w && w._node) {
+                            obj.shard = w._node.pnode;
+                        }
+
                         cb(null, obj);
                     }
                 });
@@ -426,6 +434,11 @@ function getMetadata(req, res, next) {
                             metadata: md,
                             etag: (w || {})._etag
                         };
+
+                        if (w && w._node) {
+                            obj.shard = w._node.pnode;
+                        }
+
                         cb(null, obj);
                     }
                 });
@@ -449,10 +462,14 @@ function getMetadata(req, res, next) {
                     var d = new Date(r.metadata.mtime);
                     res.set('Last-Modified', d);
                 }
+
+                req.objectShard = r.shard;  // useful for logging
                 break;
 
             case 'parent':
                 req.parentMetadata = r.metadata;
+
+                req.parentShard = r.shard;  // useful for logging
                 break;
 
             default:
@@ -469,6 +486,13 @@ function getMetadata(req, res, next) {
 }
 
 
+/*
+ * Helper for getMetadata that looks up a key in Moray and fetches any roles
+ * on the object.
+ *
+ * If the key doesn't exist, it passes a dummy metadata object to the callback
+ * that can be used across various handlers.
+ */
 function loadMetadata(req, opts, callback) {
     req.moray.getMetadata(opts, function (err, md, wrap) {
         if (err) {
@@ -481,13 +505,6 @@ function loadMetadata(req, opts, callback) {
             } else {
                 return (callback(err, req));
             }
-        } else {
-            /*
-             * We want to save which shard the metadata was fetched from for
-             * logging purposes, but `wrap` will only contain this information
-             * if there wasn't an error, whether fatal or not.
-             */
-            req.shard = wrap._node.pnode;
         }
 
         if (md.roles) {
